### DISCUSSION TITLE
Strongswan not working with Velocloud SDWan after rekey

### DISCUSSION THREAD

**Discussion by vishal1kumar** [strongswan/strongswan#576] [Q&A]
Please refer [link ](https://github.com/strongswan/strongswan/discussions/576#discussioncomment-1276489)for full logs.

I am testing Strongswan-5.8.4 (installed as responder) with a Velocloud SDWan (IPSec ikev2 tunnel initiator). 
- Veloclud is configured with 300sec of IPsec SA Lifetime. 
- Strongswan is configured with 7200sec of child sa rekey time.

With above configuration, Charon is not recognizing the Rekey request coming from Velo properly and treating it as new ChildSA request:
```
(from charon.log): <a.b.c.e> is the velo egress ip:
 1092021 MDT 00:08:53.389 12[NET] <1198689922_31Aug_Sr_Velo_Test_51115|2> received packet: from <a.b.c.e>[4500] to 172.17.0.11[4500] (348 bytes)
 1092021 MDT 00:08:53.389 12[ENC] <1198689922_31Aug_Sr_Velo_Test_51115|2> parsed CREATE_CHILD_SA request 13 [ SA No KE TSi TSr ]
 1092021 MDT 00:08:53.389 12[CFG] <1198689922_31Aug_Sr_Velo_Test_51115|2> selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/MODP_1024/NO_EXT_SEQ
 1092021 MDT 00:08:53.390 12[CFG] <1198689922_31Aug_Sr_Velo_Test_51115|2> unable to install policy 172.0.0.0/8 === 0.0.0.0/0 in (mark 51115/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 1092021 MDT 00:08:53.390 12[CFG] <1198689922_31Aug_Sr_Velo_Test_51115|2> unable to install policy 172.0.0.0/8 === 0.0.0.0/0 fwd (mark 51115/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 1092021 MDT 00:08:53.390 12[CFG] <1198689922_31Aug_Sr_Velo_Test_51115|2> unable to install policy 0.0.0.0/0 === 172.0.0.0/8 out (mark 51115/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 1092021 MDT 00:08:53.390 12[IKE] <1198689922_31Aug_Sr_Velo_Test_51115|2> unable to install IPsec policies (SPD) in kernel
 1092021 MDT 00:08:53.390 12[IKE] <1198689922_31Aug_Sr_Velo_Test_51115|2> failed to establish CHILD_SA, keeping IKE_SA
 1092021 MDT 00:08:53.390 12[ENC] <1198689922_31Aug_Sr_Velo_Test_51115|2> generating CREATE_CHILD_SA response 13 [ N(TS_UNACCEPT) ]
 1092021 MDT 00:08:53.390 12[NET] <1198689922_31Aug_Sr_Velo_Test_51115|2> sending packet: from 172.17.0.11[4500] to <a.b.c.e>[4500] (76 bytes)
```
After this the childSA gets deleted after some retry and never comes up. Work around to this is to keep rekey time in strongswan less than velocloud so that Strongswan initiates the rekey 1st and refreshes the SA.

If we replace the initiator with a Strongswan client it works fine
```
(from charon.log):  with a strongswan initiator:
 2092021 MDT 02:36:50.197 08[NET] <3412721420_ipsec_loc_1_50001|268> received packet: from 10.140.222.126[4500] to 172.17.0.7[4500] (316 bytes)
 2092021 MDT 02:36:50.198 08[ENC] <3412721420_ipsec_loc_1_50001|268> parsed CREATE_CHILD_SA request 2 [ N(REKEY_SA) SA No TSi TSr ]
 2092021 MDT 02:36:50.198 08[CFG] <3412721420_ipsec_loc_1_50001|268> selected proposal: ESP:AES_CBC_128/HMAC_SHA2_256_128/NO_EXT_SEQ
 2092021 MDT 02:36:50.201 08[IKE] <3412721420_ipsec_loc_1_50001|268> inbound CHILD_SA child-sa{524} established with SPIs cd7824ec_i c4f10204_o and TS 0.0.0.0/0[tcp/ftp] 0.0.0.0/0[tcp/http] 0.0.0.0/0[tcp/https] === 10.140.222.0/24
 2092021 MDT 02:36:50.201 08[ENC] <3412721420_ipsec_loc_1_50001|268> generating CREATE_CHILD_SA response 2 [ SA No TSi TSr ]
 2092021 MDT 02:36:50.202 08[NET] <3412721420_ipsec_loc_1_50001|268> sending packet: from 172.17.0.7[4500] to 10.140.222.126[4500] (236 bytes)
 2092021 MDT 02:36:50.205 09[NET] <3412721420_ipsec_loc_1_50001|268> received packet: from 10.140.222.126[4500] to 172.17.0.7[4500] (76 bytes)
 2092021 MDT 02:36:50.205 09[ENC] <3412721420_ipsec_loc_1_50001|268> parsed INFORMATIONAL request 3 [ D ]
 2092021 MDT 02:36:50.205 09[IKE] <3412721420_ipsec_loc_1_50001|268> received DELETE for ESP CHILD_SA with SPI c866a40c
 2092021 MDT 02:36:50.206 09[IKE] <3412721420_ipsec_loc_1_50001|268> closing CHILD_SA child-sa{523} with SPIs c4c10703_i (0 bytes) c866a40c_o (0 bytes) and TS 0.0.0.0/0[tcp/ftp] 0.0.0.0/0[tcp/http] 0.0.0.0/0[tcp/https] === 10.140.222.0/24
 2092021 MDT 02:36:50.207 09[IKE] <3412721420_ipsec_loc_1_50001|268> sending DELETE for ESP CHILD_SA with SPI c4c10703
 2092021 MDT 02:36:50.207 09[IKE] <3412721420_ipsec_loc_1_50001|268> CHILD_SA closed
 2092021 MDT 02:36:50.208 09[IKE] <3412721420_ipsec_loc_1_50001|268> outbound CHILD_SA child-sa{524} established with SPIs cd7824ec_i c4f10204_o and TS 0.0.0.0/0[tcp/ftp] 0.0.0.0/0[tcp/http] 0.0.0.0/0[tcp/https] === 10.140.222.0/24
 2092021 MDT 02:36:50.208 09[ENC] <3412721420_ipsec_loc_1_50001|268> generating INFORMATIONAL response 3 [ D ]
 2092021 MDT 02:36:50.208 09[NET] <3412721420_ipsec_loc_1_50001|268> sending packet: from 172.17.0.7[4500] to 10.140.222.126[4500] (76 bytes)
```
**Questions:
Where is it going wrong with Velocloud SDWan?
Is Charon not generic enough to understand the Rekey Child SA or Velocloud SDWan is not sending the Rekey request properly?**

Please suggest..

TIA

**Comment by tobiasbrunner** [comment]
Please post the complete logs and configs (see #196).

**Comment by vishal1kumar** [comment]
Sure Tobias. My bad that I missed to attach them earlier.

New logs attached at [link](https://github.com/strongswan/strongswan/discussions/576#discussioncomment-1276489)

<img width="645" alt="velo_config" src="https://user-images.githubusercontent.com/32055144/131824741-bb0fdbe2-01f1-4d30-9302-9f347c947b88.png">

**Reply by Thermi** [reply]
There seems to be a lot going wrong on your box.

Please provide complete logs. Do not remove any lines. Use the verbatim log settings from #196.

**Comment by vishal1kumar** [comment]
Hi Tobias and Thermi,
Sorry for the delay, I had to prepare the setup and run the test case again. I have collected all the logs as suggested in [#196](https://github.com/strongswan/strongswan/discussions/196).

**Test Scenario**:
- Strongswan-5.8.4 installed as responder
- Velocloud SDWan configured as IPSec ikev2 tunnel initiator
- Veloclud is configured with 300sec of IPsec SA Lifetime.
- Strongswan is configured with 7200sec of child sa rekey time.

**Observation**:
- 04:31:19: Tunnel established between Strangswan and Velo
- 04:35:20: Velocloud sends Rekey request to charon. Which charon treats as a new NEW CHILD SA creation request and replies back 'TS_UNACCEPT' to velo.
- 04:36:18: Velocloud reties every second for 1 min to try to rekey with charon.
- 04:36:19: Velocloud gives up and sends 'DELETE for ESP CHILD_SA'
- 04:36:24: Velocloud starts sending CHILS SA Create request every second for 'TS 0.0.0.0/0[0/ipsec-nat-t]'
- Charon starts creating 100s of Child SA for TS 0.0.0.0/0[0/ipsec-nat-t]. And swanctl is flooded with these duplicate sas.

**Issues**:
- charon should have acknowledge 'CHILD SA' request (received at '04:35:20') as Rekey request and processed it accordinly, rather it is treating it as new CHILD.
- In any case, charon should not allow to create duplicate child sas in error case or if initiator start flooding it with new dup TS requests.

**Logs**:
Below are the logs in the zipped file. Would really appreciate if you can have a looks and tell me what's going wrong:
[Logs_for_stongswan_discussion.zip](https://github.com/strongswan/strongswan/files/7105324/Logs_for_stongswan_discussion.zip)

Hope to listen from you soon.

TIA

**Reply by Thermi** [reply]
The first CHILD_SA request is for the following:
 3092021 MDT 04:31:19.380 09[IKE] <failingConnection|2> CHILD_SA child-sa{1} established with SPIs cd4274ba_i c0992d6a_o and TS 0.0.0.0/0 === 22.22.22.22/32 54.0.0.0/8 172.0.0.0/8

Then it follows up with one for the following:
3092021 MDT 04:35:19.717 04[CFG] <failingConnection|2> looking for a child config for 0.0.0.0/0 === 22.22.22.22/32

Installing that fails because such a selctor already exists for child-sa{1}, so it fails:
```
3092021 MDT 04:35:19.718 04[CFG] <failingConnection|2> unable to install policy 22.22.22.22/32 === 0.0.0.0/0 in (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:19.718 04[CFG] <failingConnection|2> unable to install policy 22.22.22.22/32 === 0.0.0.0/0 fwd (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:19.718 04[CFG] <failingConnection|2> unable to install policy 0.0.0.0/0 === 22.22.22.22/32 out (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:19.718 04[IKE] <failingConnection|2> unable to install IPsec policies (SPD) in kernel
 3092021 MDT 04:35:19.718 04[IKE] <failingConnection|2> failed to establish CHILD_SA, keeping IKE_SA
 ```

It tries again:
`3092021 MDT 04:35:20.718 06[CFG] <failingConnection|2> looking for a child config for 0.0.0.0/0 === 22.22.22.22/32`
and fails for the same reason as before:
```
 3092021 MDT 04:35:20.719 06[CFG] <failingConnection|2> unable to install policy 22.22.22.22/32 === 0.0.0.0/0 in (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:20.719 06[CFG] <failingConnection|2> unable to install policy 22.22.22.22/32 === 0.0.0.0/0 fwd (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:20.719 06[CFG] <failingConnection|2> unable to install policy 0.0.0.0/0 === 22.22.22.22/32 out (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists
 3092021 MDT 04:35:20.719 06[IKE] <failingConnection|2> unable to install IPsec policies (SPD) in kernel
 3092021 MDT 04:35:20.719 06[IKE] <failingConnection|2> failed to establish CHILD_SA, keeping IKE_SA
```

This happens 61 more times.

Then it requests a CHILD_SA for 
`3092021 MDT 04:36:24.142 12[CFG] <failingConnection|2> looking for a child config for 0.0.0.0/0[0/ipsec-nat-t] === 0.0.0.0/0`

That TS makes no sense: 0.0.0.0/0[0/ipsec-nat-t] === 0.0.0.0/0 (because protocol number in the local selector is 0, and the port is 4500, which needs to go outside of the tunnel in any case, if the sender, or receiver, are one of the peers.

Because your config does not cover the whole 0.0.0.0/0 for the remote peer, it is reduced:
```
3092021 MDT 04:36:24.810 10[CFG] <failingConnection|2> selecting traffic selectors for us:
 3092021 MDT 04:36:24.813 10[CFG] <failingConnection|2>  config: 0.0.0.0/0, received: 0.0.0.0/0[0/ipsec-nat-t] => match: 0.0.0.0/0[0/ipsec-nat-t]
 3092021 MDT 04:36:24.813 10[CFG] <failingConnection|2> selecting traffic selectors for other:
 3092021 MDT 04:36:24.813 10[CFG] <failingConnection|2>  config: 22.22.22.22/32, received: 0.0.0.0/0 => match: 22.22.22.22/32
 3092021 MDT 04:36:24.813 10[CFG] <failingConnection|2>  config: 54.0.0.0/8, received: 0.0.0.0/0 => match: 54.0.0.0/8
 3092021 MDT 04:36:24.813 10[CFG] <failingConnection|2>  config: 172.0.0.0/8, received: 0.0.0.0/0 => match: 172.0.0.0/8
 3092021 MDT 04:36:24.813 10[CHD] <failingConnection|2> CHILD_SA child-sa{63} state change: CREATED => INSTALLING
```

It then requests the same TS again, 52 times.

That's definitely a failure on part of the VeloCloud SDWan appliance.
It should..
1) Stop requesting such a CHILD_SA  once it does not get the TS it requests and notify the admin/log an error/whatever
2) Destroy the old CHILD_SAs when it tries to create a new one

I guess you could try working around the issue by configuring remote_ts to be 0.0.0.0/0. Maybe that stops the last problem from occuring.

**Reply by tobiasbrunner** [reply]
The other box seems to have trouble with the traffic selector (TS) narrowing done by strongSwan. Since you configured three TS in `remote_ts`, the proposed TS `0.0.0.0/0` is narrowed:

    3092021 MDT 04:31:19.379 09[CFG] <failingConnection|2> selecting traffic selectors for other:
    3092021 MDT 04:31:19.379 09[CFG] <failingConnection|2>  config: 22.22.22.22/32, received: 0.0.0.0/0 => match: 22.22.22.22/32
    3092021 MDT 04:31:19.379 09[CFG] <failingConnection|2>  config: 54.0.0.0/8, received: 0.0.0.0/0 => match: 54.0.0.0/8
    3092021 MDT 04:31:19.379 09[CFG] <failingConnection|2>  config: 172.0.0.0/8, received: 0.0.0.0/0 => match: 172.0.0.0/8
    ...
    3092021 MDT 04:31:19.380 09[IKE] <failingConnection|2> CHILD_SA child-sa{1} established with SPIs cd4274ba_i c0992d6a_o and TS 0.0.0.0/0 === 22.22.22.22/32 54.0.0.0/8 172.0.0.0/8

This seems to get accepted by the peer. However, did you actually test if traffic to all three TS flows?

Anyway, where it gets weird is during the "rekeying", which it actually isn't. Instead of properly rekeying the existing CHILD_SA, the peer just sends a request for a new CHILD_SA:

    3092021 MDT 04:35:19.716 04[ENC] <failingConnection|2> parsed CREATE_CHILD_SA request 13 [ SA No KE TSi TSr ]

Proper CHILD_SA rekeyings contain a `REKEY_SA` notify, referencing the SPI of the previous SA - maybe this is one of those crappy IKEv2 implementations that are hacked together from an existing IKEv1 implementation.  The latter assumption is kinda reinforced by the next observation.

Instead of proposing `0.0.0.0/0`, which it used before, the other box now actually only uses the first of the previous TS (IKEv1 only supported negotiating a single TS per CHILD_SA, so that might be why):

    3092021 MDT 04:35:19.717 04[CFG] <failingConnection|2> selecting traffic selectors for other:
    3092021 MDT 04:35:19.717 04[CFG] <failingConnection|2>  config: 22.22.22.22/32, received: 22.22.22.22/32 => match: 22.22.22.22/32
    3092021 MDT 04:35:19.717 04[CFG] <failingConnection|2>  config: 54.0.0.0/8, received: 22.22.22.22/32 => no match
    3092021 MDT 04:35:19.717 04[CFG] <failingConnection|2>  config: 172.0.0.0/8, received: 22.22.22.22/32 => no match

This results in a new set of traffic selectors for which a new reqid is allocated, but the first IPsec policy that's derived from it (using the first local and remote TS) actually conflicts with one of the existing policies of the previously negotiated CHILD_SA:

    3092021 MDT 04:35:19.718 04[CFG] <failingConnection|2> unable to install policy 22.22.22.22/32 === 0.0.0.0/0 in (mark 51116/0xffffffff) for reqid 2, the same policy for reqid 1 exists

It then repeatedly tries to create the same CHILD_SA but never succeeds. Eventually, it deletes the initial CHILD_SA (maybe due to a hard expire).

Now comes this very weird request with `0.0.0.0/0[0/ipsec-nat-t]` as TS, which is neither `0.0.0.0/0` nor `22.22.22.22/32` and makes absolutely no sense.

And it doesn't stop there. Due to the configured `local_ts=0.0.0.0/0`, this proposal is actually accepted. But the peer doesn't seem happy about it and now just continues to send requests for CHILD_SAs with that identical unusable TS.

Looking at the image you posted above, I wonder if you actually configured multiple subnets in the "Site Subnets" list (or only one as the image suggests), and whether the options "Disable Site Subnets" and "Enable Tunnel(s)" make a difference. For a route based approach, you might want to configure `remote_ts=0.0.0.0/0` instead of the actual remote subnets (it's what the peer proposes anyway).

**Reply by vishal1kumar** [reply]
Thanks Tobias and Thermi for your analysis. I see you have identified two issues with the Velo SDWan:
- Velocloud's Rekey request if missing 'REKEY_SA' notification referencing the SPI of the previous SA.
- After hard expire of rekey time, velo starts sending unexpected SA request for [0/ipsec-nat-t] in a non-stop mode.

We are planning to initiate a discussion thread with Velo to report/investigate these issues. I will update this thread once I have a response from them.

Thanks
